### ISSUE TITLE
Sending WebSocket ping

### ISSUE THREAD

**Issue by francisdb** [playframework/playframework#3861]
I was looking for a way to send ping websocket frames with play.

I see the server properly handles incoming ping frames as this was fixed here: https://github.com/playframework/playframework/pull/2130

According to [mozilla](https://developer.mozilla.org/en-US/docs/WebSockets/Writing_WebSocket_servers#Pings_and_Pongs.3A_The_Heartbeat_of_WebSockets) "either the client or the server can choose to send a Ping to the other party" so it makes sense to have this available on serverside

**Comment by jroper** [comment]
Yep, this isn't supported, we do have plans to allow it in future though, probably won't be till at least Play 3.

**Comment by mohamedamineniso** [comment]
I will participate on the issue !

**Comment by rabzu** [comment]
+1

**Comment by folex** [comment]
Hi! This is still not implemented? 
I can't send pings from javascript websocket, so I need to send pings from server.

**Comment by francisdb** [comment]
There is a thread on the akka http mailing list related to this
https://groups.google.com/forum/#!topic/akka-user/GCim2YNa7w8

And a ticket for serverside ping support
https://github.com/akka/akka-http/issues/102

**Comment by ktoso** [comment]
This would be best implemented on the Akka side, and Play make use of that.
Help is very welcome on this, feel free to help out by doing a PR!  akka/akka-http#102

**Comment by CEikermann** [comment]
Akka have implemented the automatic ping for websockets:
https://github.com/akka/akka-http/issues/102

Can we already set this up within the Play Framework (e.g. automatic ping send once per minute)?

**Comment by dope9967** [comment]
Implemented with Akka HTTP 10.1.1 and it's version was updated in PlayFramework with commit https://github.com/playframework/playframework/commit/20add4ca4000ead7791c23e98fdc1f09b6465892, so release [2.6.13](https://github.com/playframework/playframework/releases/tag/2.6.13) already has configurable server-side WebSocket ping. Issue should probably be closed now.

**Comment by dope9967** [comment]
Actually I was wrong - Akka HTTP 10.1.1 was merged to Play Framework master, but not 2.6.x branch. Seems like 10.1.x is not planed for 2.6.x. I there anything holding back update to 10.1.1 for 2.6.x branch?

**Comment by gmethvin** [comment]
There are some compatibility breakages in 10.1.1: https://doc.akka.io/docs/akka-http/current/migration-guide/migration-guide-10.1.x.html. It should be safe to upgrade in your own code if you follow the migration guide, but it would be a bad idea for Play to force the upgrade on users.

**Comment by lvxudong** [comment]
@gmethvin Hi, Can i know whether if it was solved in latest Play 2.7 release?

**Comment by marcospereira** [comment]
> Can i know whether if it was solved in latest Play 2.7 release?

It was not resolved in Play 2.7, @lvxudong.

**Comment by atropo** [comment]
Is this solved in Play 2.8 release?

I cannot make it work using the setting in `application.conf`.

**Comment by mkurz** [comment]
Actually that _should_ work in Play 2.8.8 because it uses latest akka-http 10.1.x.
However I am not sure if we need to enable the ping support in Play explicitly somehow.
@atropo Can you please try setting, in your `build.sbt`:
```
PlayKeys.devSettings += "akka.http.server.websocket.periodic-keep-alive-max-idle" -> "1 second"
```
and in your `application.conf`:
```
akka.http.server.websocket.periodic-keep-alive-max-idle = 1 second
```
?
Are you using Play 2.8.8?

**Comment by ocordsen** [comment]
@mkurz Do you have any other hint for me or should I try to find another solution for the next time? If it would help, I could prepare a minimal example repository. Thanks for your help again.

**Comment by mkurz** [comment]
@ocordsen sorry no other hint right now (I am low on time as well). A reproducer would probably help.

**Comment by ocordsen** [comment]
@mkurz I created a example project: https://github.com/ocordsen/play-framework-web-socket-keep-alive-ping
I tested in dev mode via `sbt run` and in prod mode. The view returned on the route `/` of the project connects to the websocket and logs Connect and Disconnect (in JS console). I cannot see any ping frames in the chrome dev tools network tab and every 75 seconds I have a reconnect.

**Comment by ocordsen** [comment]
Additional info: I tried to implement a custom AkkaHttpServer (as described here: https://www.playframework.com/documentation/2.8.x/SettingsAkkaHttp#Direct-Akka-HTTP-configuration) which was successfully loaded by my application but the keep-alive setting didn't work with this approach, too.

**Comment by joesan** [comment]
Is there any update on this? Which version of the Play framework has this support?

**Comment by joesan** [comment]
@mkurz Which example should I look at? I suppose it should be this one https://github.com/playframework/play-samples/tree/2.8.x/play-scala-websocket-example

Could you please update a demo of how that works to keep-alive websocket connections? Is this a workaround as mentioned in the Play Framework documentation?

**Comment by joesan** [comment]
I have a workaround solution implemented where I spin up a stream that emits a value at regular intervals and I pass the emitted value to the Actor instance which then simply maps that to a HeartBeat message back to the client. This way, I can have a persistent connection enabled. Sort of ugly, but it works.

**Comment by joesan** [comment]
Here is a solution that I tried.

https://stackoverflow.com/questions/73904123/play-framework-persistent-websocket-connection

Is that good enough?

**Comment by mkurz** [comment]
I implemented a patch to finally make this a reality: #11521
I am planning to release that in Play 2.8.19.

**Comment by mkurz** [comment]
This has been no been fixed in #11521 and will make its way into Play 2.8.19. Please have a look at the docs I added in that PR (also see [the screenshot](https://github.com/playframework/playframework/pull/11521#issuecomment-1298381513) :wink:)
